src/uuid/uuidbuilder.rs

```diff
@@ -33,174 +32,171 @@ pub struct UUIDBuilder {
+
+        // Obtain a UUID before spawning tasks to determine the expected LSB and ensure consistent
+        // across all threads / tasks
+        let expected_lsb = UUIDBuilder::build().lsb;
+
+        let mut tasks = Vec::new();
+
+        // Record the start time before any tasks are spawned
+        let start = Instant::now();
+
+        for task_id in 0..num_tasks {
+            let expected_lsb_clone = expected_lsb;
+            let task = task::spawn(async move {
+                let mut local_uuids = Vec::new();
+                for _ in 0..uuids_per_task {
+                    let uuid = UUIDBuilder::build();
+                    assert_eq!(
+                        uuid.lsb, expected_lsb_clone,
+                        "LSB does not match the expected value."
+                    );
+                    local_uuids.push((task_id, uuid));
+                }
+                local_uuids
+            });
+            tasks.push(task);
+        }
 
-    #[test]
-    fn test_build_with_instant_creates_uprotocol_uuid() {
-        let instant = 0x18C684468F8_u64; // Thu, 14 Dec 2023 12:19:23 GMT
-        let uuid = UUIDBuilder::new().build_with_instant(instant);
-        assert!(uuid.is_uprotocol_uuid());
-        assert_eq!(uuid.get_time().unwrap(), instant);
-
-        // instant, version (8) and counter (000) should show up in UUID
-        assert!(uuid
-            .to_hyphenated_string()
-            .starts_with("018c6844-68f8-8000-"));
-    }
+        // Await all tasks and collect their results
+        let results = futures::future::join_all(tasks).await;
 
-    #[test]
-    fn test_uuid_for_subsequent_generation() {
-        let instant = 0x18C684468F8_u64; // Thu, 14 Dec 2023 12:19:23 GMT
-        let builder = UUIDBuilder::new();
-
-        let uuid_for_instant = builder.build_with_instant(instant);
-        assert!(uuid_for_instant.is_uprotocol_uuid());
-        // instant, version (8) and counter (000) should show up in UUID
-        assert!(uuid_for_instant
-            .to_hyphenated_string()
-            .starts_with("018c6844-68f8-8000-"));
-
-        let uuid_for_same_instant = builder.build_with_instant(instant);
-        assert!(uuid_for_same_instant.is_uprotocol_uuid());
-        // same instant, version (8) and _incremented_ counter (001) should show up in UUID
-        assert!(uuid_for_same_instant
-            .to_hyphenated_string()
-            .starts_with("018c6844-68f8-8001-"));
-    }
+        // Record the end time after all tasks are completed
+        let end = Instant::now();
 
-    #[test]
-    fn test_uuid_for_constant_random() {
-        let factory = UUIDBuilder::new();
-        let uuid1 = factory.build();
-        let uuid2 = factory.build();
-        assert_eq!(uuid1.lsb, uuid2.lsb);
-    }
+        let duration = end.duration_since(start);
+
+        info!("All tasks completed in {:?}.", duration);
```

Comment:
Is it really worth it/necessary to add the log dependency? I'm not against using proper logging, but does it add sufficient value in this case?

Reply:
I'm okay making it a dev dependency as it's only used in tests 👍

I suppose there's the one tricky case of if we generated more than `counter` number of UUIDs per ms that we could put a warn!() for. But realistically this likely won't occur.

What do you think?

Reply:
Well this we will have to approach in a more generic way, should this be a valid scenario - logging this info during tests will quite surely be lost on the world ;-)

Famous last words would be "we'll never need more than `counter` ids per ms!". To put this differently: do we see ourselves even approaching that boundary in the coming years? Or differently again: if we're afeared of this scenario, we might want to increase the size of counter generally, or ensure in the code that if counter is "full", we don't generate any more ids until the next ms turns around.

I'd favour removing the logging dependency and those two output lines entirely, and rather replace with asserts in some form. E.g., 'assert that we're not overrunning counter within the same ms'.

Does that make sense?

Reply:
These seem pointed at @stevenhartley :slightly_smiling_face: 

> Famous last words would be "we'll never need more than counter ids per ms!". To put this differently: do we see ourselves even approaching that boundary in the coming years? Or differently again: if we're afeared of this scenario, we might want to increase the size of counter generally, or ensure in the code that if counter is "full", we don't generate any more ids until the next ms turns around.

----------

> I'd favour removing the logging dependency and those two output lines entirely, and rather replace with asserts in some form. E.g., 'assert that we're not overrunning counter within the same ms'.
> 
> Does that make sense?

Yup, I can look into this.

Reply:
> I'd favour removing the logging dependency and those two output lines entirely, and rather replace with asserts in some form. E.g., 'assert that we're not overrunning counter within the same ms'.
> 
> Does that make sense?

Pushed up a change which:
1. Moves the logging deps over to dev dependencies
   * I think these logging lines are still valuable for the case if someone ever changes the test case to have > 10 tasks or > 409 UUIDs to be generated per task to clarify why it broke
2. Clarifies the asserts check if we've overrun the counter

Hope that resolves the issue, if not lemme know :slightly_smiling_face:

Reply:
With the change I have proposed above, this should never happen again ....